kong/tools/database_cache.lua

```diff
@@ -159,30 +165,42 @@ function _M.get_or_set(key, ttl, cb, ...)
   end
 
   -- Lock acquired. Since in the meantime another worker may have
-  -- populated the value we have to check again
+  -- populated the value we have to check again.
+  -- Errors will be collected, but only dealt with AFTER unlocking
   value = _M.get(key)
-  if value == nil then
-    -- Get from closure
-    value, err = cb(...)
-    if err then
-      return nil, err
+  if value ~= nil then
+    -- shm cache success
+    local ok, lerr = lock:unlock()
+    if not ok and lerr then
+      ngx_log(ngx.ERR, "failed to unlock: ", lerr)
     end
 
-    if value ~= nil then
-      local ok, err = _M.set(key, value, ttl)
-      if not ok then
-        ngx_log(ngx.ERR, err)
-        return
-      end
+    return value
+  end
+
+  -- cache failed, we need to invoke the callback
+  value = pack(cb(...))
+
+  if value[1] ~= nil then
+
+    local ok, err = _M.set(key, value[1], ttl)
+    if not ok then
+      ngx_log(ngx.ERR, err)
     end
   end
 
-  local ok, err = lock:unlock()
-  if not ok and err then
-    ngx_log(ngx.ERR, "failed to unlock: ", err)
+  local ok, lerr = lock:unlock()
+  if not ok and lerr then
+    ngx_log(ngx.ERR, "failed to unlock: ", lerr)
+  end
+
+  if value[1] ~= nil then
+    -- only return first value, as on each next lookup the cache will
+    -- also only return a single value
+    return value[1]
   end
 
-  return value
+  return unpack(value)
```

Comment:
A few things:

- in places where we hit a lock error, we just `return`
- not in the new places L.194 and L.188
- here on L.203, we might return `nil, nil, ...`

All of these points are a problem because the way we check for errors after `get_ot_set()` is like this:
```
local thing, err = cache.get_or_set(...)
if err then
  return responses.send_HTTP_INTERNAL_SERVER_ERROR(err)
end
```

Hence, I feel like we should always return an error to the caller if such happens in `get_or_set()`, instead of failing at a later point in the caller's code and make tracing of the error harder.

I think the main issue is the callers are not handling cache miss properly (missing entity from the DB hit), and they should, and lock issues should be reported internally only. But this is the fastest change we can do to address it right now.